conftest.py

```diff
@@ -109,3 +109,12 @@ def get_mesos_state_on_failure(test_name: str):
         log_name = '{}_slaves.json'.format(test_name)
         with open(log_name, 'w') as f:
             f.write(r.text)
+
+
+def pytest_runtest_setup(item):
+    if isinstance(item, item.Function):
+        min_version_mark = item.get_marker('dcos_min_version')
+        if min_version_mark:
+            min_version = min_version_mark.args[0]
+            if sdk_utils.dcos_version_less_than(min_version):
```

Comment:
I don't know if this will work where we are going with the scripting. These plugins are loaded in first with the conftest.py files before any dependencies are actually loaded or test fixtures evaluated. Thus, this might only work now because we authenticate with the cluster outside the test run, but when we go to authenticate in ONLY the fixture, this might explode because it requires talking to a setup CLI

Reply:
Good point. If we were going to do our authentication in a fixture you're right. We could easily plan to do the authentication in a hook function as well. Possibly even right here so we auth and then kick out any invalid tests. Keeping those close together doesn't seem bad. (And I like how following the `@pytest.mark.dcos_min_version(V)` pattern keeps things pretty standard.)

I've started liking plugins more and more as we've moved through this refactoring. The idea of providing a pytest plugin to manage tests feels like a simple, _installable_ way to share a testing framework -- instead of documenting how to configure a bunch of dependent fixtures. I'm not ready to argue for that yet.... but that's where my mind has been going lately.

Reply:
A decorator version I could add to `sdk_utils.py`:

```
def dcos_min_version(f, min_version)
    @functools.wraps(f)
    def dcos_min_version_wrapper(*args):
        if dcos_version_less_than(min_version):
            pytest.skip('Feature only supported in DC/OS {} and up'.format(min_version))
        else:
            return f(*args)
    return dcos_min_version_wrapper
```

That would run after fixtures. Because it uses `skip` we should get an accurate report at the end of the run. Cluster auth could be done in a fixture or hook function. I don't like it, but I don't know why.

Reply:
I actually really like the idea of doing the auth (and perhaps getting some cluster attributes) in a plugin. I thought the plugins were a little hard to work with because its not super clear where everything fits in the pytest framework. However, for things like this, I think most devs never need to touch it, so its alright to be harder to work with.

That being said, its always easier said than done so we just gotta dive in and see...but you are about to be out for a while. For the moment,  I prefer the decorator greatly as its the same deal we have now and we can at least finish the auth-in-fixture work while you are off

Reply:
Ok. I'll get the decorator version working today, and we can revisit the plugin idea. Or you can go crazy and implement the whole thing on a bender weekend while I'm gone. 😆